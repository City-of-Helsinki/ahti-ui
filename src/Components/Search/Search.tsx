import React, { useEffect, useRef, useState } from 'react';
import styles from './Search.module.scss';
import { useTranslation } from 'react-i18next';
import mapData from '../../data/mapData.json';
import { escapeRegExp } from 'lodash';
import { useHistory } from 'react-router-dom';

import searchIcon from '../../assets/icons/search.svg';
import exitIcon from '../../assets/icons/cross.svg';

const features = mapData.features;

interface MapData {
  readonly name: string;
  readonly location: string;
  readonly type: string;
}

interface SearchItem extends MapData {
  readonly matchedPart: string;
}

const SearchItem: React.FC<SearchItem> = props => {
  const re = new RegExp(`(${escapeRegExp(props.matchedPart)})`, 'gi');
  const parts = props.name.split(re);

  let history = useHistory();

  return (
    <div
      className={styles.searchItem}
      onClick={() => history.push(`/map?name=${props.name}`)}
    >
      <img
        src={`/icons/type/${props.type}.svg`}
        alt="info image"
        style={{
          height: 30,
          width: 30,
          cursor: 'pointer',
        }}
      />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part, i) =>
            re.test(part) ? (
              <mark key={i}>{part}</mark>
            ) : (
              <span key={i}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{props.location}</div>
      </div>
    </div>
  );
};

const Search: React.FC = () => {
  const [currentSearch, setCurrentSearch] = useState<string>('');
  const { i18n } = useTranslation();
  const history = useHistory();

  const inputRef = useRef<any>(null);
  useEffect(() => {
    if (inputRef) {
      inputRef.current.focus();
    }
  });

  const data: MapData[] = features.map((feature: any) => {
    const name = feature.properties[i18n.language].name;
    const location = feature.properties.address;
    const type = feature.properties.type;
    return { name, location, type };
  });

  return (
    <div className={styles.container}>
      <div className={styles.search}>
        <img src={exitIcon} alt="close" onClick={() => history.goBack()} />
        <div className={styles.searchInputWithIcon}>
          <input
            ref={inputRef}
            type="text"
            onChange={event => setCurrentSearch(event.target.value)}
            className={styles.searchInput}
          />
          <img src={searchIcon} alt="search" />
        </div>
      </div>
      {currentSearch === '' && (
        <>
          <p>{'Viimeisimm√§t'}</p>
          <p>{'Suosittelemme'}</p>
        </>
      )}
      {currentSearch !== '' &&
        data
          .filter((place: MapData) =>
            place.name.toLowerCase().includes(currentSearch.toLowerCase())
          )
          .map((place: MapData) => (
            <SearchItem {...place} matchedPart={currentSearch} />
          ))}
    </div>
  );
};

export default Search;
