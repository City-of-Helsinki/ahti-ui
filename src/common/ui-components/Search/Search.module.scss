@import '../../../styles/icons';

.container {
  width: 100%;
  position: absolute;
  background-color: white;
  padding-left: 1rem;
  padding-right: 1rem;

  /* Hide the scrollbar, but keep scrolling enabled */
  overflow-y: scroll;

  /* Firefox */
  scrollbar-width: none;

  /* Internet Explorer 10+ */
  -ms-overflow-style: none;
}

.container::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.searchItem {
  display: flex;
  flex-direction: row;
}

.infoContainer {
  width: 100%;
  border-bottom: 1px;
  border-bottom-style: solid;
  border-bottom-color: var(--hds-ui-color-black-10);
  margin-left: 1rem;
}

.name {
  font-size: 16px;
  font-weight: 700;
  line-height: 26px;
}

.nameMark {
  all: unset;
  color: var(--hds-brand-color-coat-of-arms-blue);
}

.address {
  margin-bottom: 0.5rem;
  color: var(--hds-ui-color-black-70);
  font-weight: 400;
}

.searchInput {
  width: 100%;
  border: none;
  margin-left: 1rem;
}

.searchInput:focus {
  outline: none;
}

.searchInputWithIcon {
  display: flex;
  flex-direction: row;
  width: 100%;
  padding-bottom: 1rem;
  border-bottom: 1.5px;
  border-bottom-style: solid;
}

.infoImage {
  height: 30px;
  width: 30px;
  cursor: pointer;
}

.search {
  display: flex;
  flex-direction: row;
  margin-top: 1rem;
  margin-bottom: 1rem;
}
