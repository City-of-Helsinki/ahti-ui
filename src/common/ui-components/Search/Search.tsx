import React, { useEffect, useState } from 'react';
import styles from './Search.module.scss';
import { IconClose, IconSearch } from 'hds-react';
import {
  FeatureProperties,
  useFeaturesSearchQuery
} from '../../../domain/api/generated/types.d';
import { IconLocation } from 'hds-react';
import escapeRegExp from './utils/escapeRegExp';
import classNames from 'classnames';

export interface SearchData {
  readonly id: string;
  readonly name: string;
  readonly location?: string;
  readonly category?: string;
}

interface CategoryIconProps {
  readonly category?: string;
}

const CategoryIcon: React.FC<CategoryIconProps> = ({ category }) => {
  // BE only has ahti:category:island for the time being
  // other categories are null.
  if (category === 'ahti:category:island') {
    return <IconLocation className={styles.bigIcon} />;
  } else {
    return <IconLocation className={styles.bigIcon} />;
  }
};

interface SearchItemProps extends SearchData {
  readonly currentSearch: string;
  onSelect(id: string): void;
}

export const SearchItem: React.FC<SearchItemProps> = ({
  id,
  name,
  location,
  category,
  currentSearch,
  onSelect
}) => {
  const re = new RegExp(`(${escapeRegExp(currentSearch)})`, 'gi');
  const parts = name.split(re);

  return (
    <div className={styles.searchItem} onClick={() => onSelect(id)}>
      <CategoryIcon category={category} />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part: string, id: number) =>
            re.test(part) ? (
              <mark key={id} className={styles.nameMark}>
                {part}
              </mark>
            ) : (
              <span key={id}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{location}</div>
      </div>
    </div>
  );
};

interface SearchProps {
  readonly className?: string;
  readonly maxItems?: number;
  onSelect(id: string): void;
}

const Search: React.FC<SearchProps> = ({
  className,
  maxItems = 20,
  onSelect
}) => {
  const { data } = useFeaturesSearchQuery();
  const [searchResults, setSearchResults] = useState<SearchData[]>([]);
  const [currentSearch, setCurrentSearch] = useState<string>('');

  useEffect(() => {
    if (data && data.features && currentSearch !== '') {
      const properties = data.features.edges.reduce((acc, edge) => {
        if (edge && edge.node && edge.node.properties) {
          return [...acc, edge.node.properties];
        }
        return acc;
      }, []);
      setSearchResults(
        properties
          .map((property: FeatureProperties) => {
            return {
              id: property.ahtiId,
              name: property.name,
              location: property?.contactInfo?.address?.municipality,
              category: property.category?.id
            };
          })
          .filter((searchResult: SearchData) =>
            searchResult.name
              .toLowerCase()
              .includes(currentSearch.toLowerCase())
          )
          .slice(0, maxItems)
      );
    } else {
      setSearchResults([]);
    }
  }, [currentSearch, data]);

  return (
    <div className={classNames(styles.container, className)}>
      <div className={styles.search}>
        <div className={styles.searchInputWithIcon}>
          <IconSearch className={styles.bigIcon} />
          <input
            type="text"
            value={currentSearch}
            className={styles.searchInput}
            onChange={event => setCurrentSearch(event.target.value)}
          />
          <button aria-label="close" onClick={() => setCurrentSearch('')}>
            <IconClose className={styles.bigIcon} />
          </button>
        </div>
      </div>
      {searchResults.map((item: SearchData, id: number) => (
        <SearchItem
          key={id}
          {...item}
          currentSearch={currentSearch}
          onSelect={onSelect}
        />
      ))}
    </div>
  );
};

export default Search;
