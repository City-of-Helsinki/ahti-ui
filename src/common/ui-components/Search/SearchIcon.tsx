import React from 'react';
import classNames from 'classnames';

import styles from './Search.module.scss';

const SearchIcon: React.FC<any> = ({}) => {
  return (
    <div className={classNames(styles.searchIcon)}>
      <span></span>
      <span></span>
      <span></span>
      <span></span>
    </div>
  );
};

export default SearchIcon;
