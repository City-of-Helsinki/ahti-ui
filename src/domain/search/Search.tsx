import React, { useEffect, useRef, useState } from 'react';
import styles from './Search.module.scss';
import { useTranslation } from 'react-i18next';
import mapData from '../../data/mapData.json';
import { escapeRegExp } from 'lodash';
import { useHistory } from 'react-router-dom';

import searchIcon from '../../assets/icons/search.svg';
import exitIcon from '../../assets/icons/cross.svg';

const features = mapData.features;

export interface SearchData {
  readonly name: string;
  readonly location: string;
  readonly type: string;
}

interface SearchItemProps extends SearchData {
  readonly matchedPart: string;
  onSelect(item: string): void;
}

export const SearchItem: React.FC<SearchItemProps> = props => {
  const re = new RegExp(`(${escapeRegExp(props.matchedPart)})`, 'gi');
  const parts = props.name.split(re);

  return (
    <div
      className={styles.searchItem}
      onClick={() => props.onSelect(props.name)}
    >
      <img
        src={`/icons/type/${props.type}.svg`}
        alt="info image"
        className={styles.infoImage}
      />
      <div className={styles.infoContainer}>
        <div className={styles.name}>
          {parts.map((part, i) =>
            re.test(part) ? (
              <mark key={i}>{part}</mark>
            ) : (
              <span key={i}>{part}</span>
            )
          )}
        </div>
        <div className={styles.address}>{props.location}</div>
      </div>
    </div>
  );
};

interface SearchProps {
  readonly data: SearchData[];
  onClose(): void;
  onSelect(item: string): void;
}

const Search: React.FC<SearchProps> = ({ data, onClose, onSelect }) => {
  const [currentSearch, setCurrentSearch] = useState<string>('');
  const inputRef = useRef<any>(null);

  useEffect(() => {
    if (inputRef) {
      inputRef.current.focus();
    }
  });

  return (
    <div className={styles.container}>
      <div className={styles.search}>
        <img src={exitIcon} alt="close" onClick={() => onClose()} />
        <div className={styles.searchInputWithIcon}>
          <input
            ref={inputRef}
            type="text"
            onChange={event => setCurrentSearch(event.target.value)}
            className={styles.searchInput}
          />
          <img src={searchIcon} alt="search" />
        </div>
      </div>
      {currentSearch === '' && (
        <>
          <p>{'Viimeisimm√§t'}</p>
          <p>{'Suosittelemme'}</p>
        </>
      )}
      {currentSearch !== '' &&
        data
          .filter((item: SearchData) =>
            item.name.toLowerCase().includes(currentSearch.toLowerCase())
          )
          .map((item: SearchData, id: number) => (
            <SearchItem
              key={id}
              {...item}
              matchedPart={currentSearch}
              onSelect={onSelect}
            />
          ))}
    </div>
  );
};

export const SearchWithHistoryHook: React.FC = () => {
  const history = useHistory();
  const { i18n } = useTranslation();

  const data: SearchData[] = features.map((feature: any) => {
    const name = feature.properties[i18n.language].name;
    const location = feature.properties.address;
    const type = feature.properties.type;
    return { name, location, type };
  });
  return (
    <Search
      data={data}
      onClose={() => history.goBack()}
      onSelect={name => history.push(`/map?name=${name}`)}
    />
  );
};

export default Search;
