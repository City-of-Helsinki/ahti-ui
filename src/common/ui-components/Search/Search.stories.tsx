import React from 'react';
import Search from './Search';
import { ApolloProvider } from '@apollo/react-common';
import graphQLClient from '../../../domain/api';

export default {
  title: 'Search',
  component: Search,
  decorators: [
    (storyFn: () => React.ReactNode) => (
      <ApolloProvider client={graphQLClient}>{storyFn()}</ApolloProvider>
    )
  ]
};

export const DefaultSearch = () => (
  <Search
    onSelect={item => {
      alert(`clicked: ${item}`);
    }}
  />
);

DefaultSearch.story = {
  name: 'Default Search'
};
