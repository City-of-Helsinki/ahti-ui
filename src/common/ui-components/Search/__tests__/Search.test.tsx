import React from 'react';
import Search, { SearchData, SearchItem } from '../Search';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';
import waitForExpect from 'wait-for-expect';
import FEATURES_SEARCH_QUERY from '../../../../domain/api/queries/featuresSearchQuery';

describe('Search', () => {
  let searchWrapper: ReactWrapper;
  const mocks = [
    {
      request: { query: FEATURES_SEARCH_QUERY },
      result: {
        features: {
          edges: [
            {
              node: {
                properties: {
                  ahtiId: 'myhelsinki:place:364',
                  category: {
                    id: 'ahti:category:island'
                  },
                  name: 'Harakan luontokeskus',
                  contactInfo: {
                    address: {
                      municipality: 'Helsinki'
                    }
                  }
                }
              }
            }
          ]
        }
      }
    }
  ];

  beforeEach(() => {
    searchWrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Search onSelect={jest.fn()} />
      </MockedProvider>
    );
  });

  it('matches snapshot', () => {
    expect(searchWrapper.html()).toMatchSnapshot();
  });

  it('returns items on search', async () => {
    expect(searchWrapper.find(SearchItem)).toHaveLength(0);
    searchWrapper.find('input').simulate('change', { target: { value: 'a' } });
    await act(async () => {
      waitForExpect(() => {
        expect(searchWrapper.find(SearchItem)).toHaveLength(1);
      });
    });
  });

  it('clicking cross clears current search', () => {
    searchWrapper.find('input').simulate('change', { target: { value: 'a' } });
    searchWrapper.find('button[aria-label="close"]').simulate('click');
    expect(searchWrapper.find('input').prop('value')).toEqual('');
  });
});

const mockData: SearchData[] = [
  {
    id: 'myhelsinki:place:1',
    name: 'Venepaikka',
    location: 'Helsinki',
    category: 'ahti:category:island'
  }
];

describe('SearchItem', () => {
  let searchItemWrapper: ShallowWrapper;
  let mockOnSelect: jest.Mock;

  beforeEach(() => {
    mockOnSelect = jest.fn();
    searchItemWrapper = shallow(
      <SearchItem
        currentSearch={'Vene'}
        onSelect={mockOnSelect}
        {...mockData[0]}
      />
    );
  });

  it('matches snapshot', () => {
    expect(searchItemWrapper.html()).toMatchSnapshot();
  });

  it('calls onSelect on item click', () => {
    expect(mockOnSelect).toHaveBeenCalledTimes(0);
    searchItemWrapper.simulate('click');
    expect(mockOnSelect).toHaveBeenCalledTimes(1);
    expect(mockOnSelect).toHaveBeenCalledWith('myhelsinki:place:1');
  });
});
