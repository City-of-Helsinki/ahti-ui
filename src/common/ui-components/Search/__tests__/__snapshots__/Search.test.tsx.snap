// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search matches snapshot 1`] = `"<div class=\\"container\\"><div class=\\"search\\"><div class=\\"searchInputWithIcon\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 24 24\\" class=\\"bigIcon\\"><path d=\\"M15.3582734,0.0535251799 C12.0567815,0.0517959979 9.04784082,1.94662012 7.62269829,4.92467659 C6.19755575,7.90273306 6.6096026,11.4345925 8.68201439,14.0046043 L0.00920863309,22.676259 L1.30820144,23.9758273 L9.98273381,15.3035971 C12.9168981,17.6690044 17.0515408,17.841428 20.1724602,15.7285309 C23.2933795,13.6156338 24.6689728,9.71271903 23.5625765,6.10989254 C22.4561801,2.50706606 19.1271543,0.0488985453 15.3582734,0.0517959979 L15.3582734,0.0535251799 Z M15.3582734,15.3657554 C11.637384,15.3657554 8.62100719,12.3493786 8.62100719,8.62848921 C8.62100719,4.90759984 11.637384,1.89122302 15.3582734,1.89122302 C19.0791628,1.89122302 22.0955396,4.90759984 22.0955396,8.62848921 C22.0917339,12.3479875 19.0777709,15.3624653 15.3582734,15.3669065 L15.3582734,15.3657554 Z\\"></path></svg><input type=\\"text\\" class=\\"searchInput\\" value=\\"\\"><button aria-label=\\"close\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 15 15\\" class=\\"bigIcon\\"><path fill-rule=\\"nonzero\\" d=\\"M13.412 0L15 1.588 9.088 7.5 15 13.412 13.412 15 7.5 9.088 1.588 15 0 13.412 5.911 7.5 0 1.588 1.588 0 7.5 5.912 13.412 0z\\"></path></svg></button></div></div></div>"`;

exports[`SearchItem matches snapshot 1`] = `"<div class=\\"searchItem\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"100%\\" height=\\"100%\\" viewBox=\\"0 0 23 23\\" class=\\"bigIcon\\"><path d=\\"M11.6911445,6.57252031e-13 C13.9380141,6.57252031e-13 16.1856662,0.848347826 17.8995793,2.54582609 C21.3274054,5.94 21.0957532,11.2296522 17.8995793,14.8382609 C14.618101,18.5446957 12.868188,20.8565217 11.6911445,23.1793043 C10.5203619,20.8565217 8.7657532,18.5446957 5.48505755,14.8382609 C2.28888364,11.2296522 2.05644886,5.94 5.48505755,2.54582609 C7.19897059,0.848347826 9.44584016,6.57252031e-13 11.6911445,6.57252031e-13 Z M11.6911445,1.73895652 C9.80818799,1.73895652 8.03792712,2.46443478 6.70827494,3.78078261 C5.39505755,5.08069565 4.71418799,6.74295652 4.73923146,8.58913043 C4.76349233,10.4126087 5.49053581,12.222 6.78653581,13.6862609 C9.03966625,16.2305217 10.568101,18.1166087 11.6927097,19.7773043 C12.8220141,18.111913 14.348101,16.2266087 16.598101,13.6862609 C17.894101,12.222 18.6211445,10.4126087 18.6454054,8.58913043 C18.6688836,6.74295652 17.9880141,5.08069565 16.6763619,3.78078261 C15.3451445,2.46443478 13.5756662,1.73895652 11.6911445,1.73895652 Z M11.6922402,5.21538261 C13.6127619,5.21538261 15.168588,6.7719913 15.168588,8.69251304 C15.168588,10.6122522 13.6127619,12.1688609 11.6922402,12.1688609 C9.77171842,12.1688609 8.21510973,10.6122522 8.21510973,8.69251304 C8.21510973,6.7719913 9.77171842,5.21538261 11.6922402,5.21538261 Z M11.6922402,6.95355652 C10.7327619,6.95355652 9.95328364,7.73381739 9.95328364,8.69251304 C9.95328364,9.65042609 10.7327619,10.430687 11.6922402,10.430687 C12.6509358,10.430687 13.4304141,9.65042609 13.4304141,8.69251304 C13.4304141,7.73381739 12.6509358,6.95355652 11.6922402,6.95355652 Z\\"></path></svg><div class=\\"infoContainer\\"><div class=\\"name\\"><span></span><mark class=\\"nameMark\\">Vene</mark><span>paikka</span></div><div class=\\"address\\">Helsinki</div></div></div>"`;
